src/emulator/workQueue.ts

```diff
@@ -0,0 +1,94 @@
+import { EmulatorLogger } from "./emulatorLogger";
+import { FunctionsExecutionMode } from "./types";
+
+type Work = () => Promise<any>;
+
+/**
+ * Queue for doing async work that can either run all work concurrently
+ * or sequentially (FIFO).
+ */
+export class WorkQueue {
+  private queue: Array<Work> = [];
+  private interval?: NodeJS.Timeout;
+  private workRunningCountInternal: number = 0;
+
+  constructor(private mode: FunctionsExecutionMode = FunctionsExecutionMode.AUTO) {}
+
+  /**
+   * Submit an entry to the queue and run it according to the WorkMode.
+   *
+   * Note: make sure start() has been called at some point.
+   */
+  submit(entry: Work) {
+    this.append(entry);
+    if (this.mode === FunctionsExecutionMode.AUTO) {
+      this.runNext();
+    }
+  }
+
+  /**
+   * Begin processing work from the queue.
+   */
+  start() {
+    if (this.mode === FunctionsExecutionMode.AUTO) {
+      return;
+    }
+
+    this.interval = setInterval(() => {
+      if (!this.workRunning) {
+        this.runNext();
+      }
+    }, 100);
```

Comment:
Huh I did not know async iterators were a thing!  Will look into this, thank you.

Reply:
Ah crap unfortunately that's Node 10+ only so we'll have to wait.

Reply:
Doesn't TypeScript transform that for you?

Reply:
If TypeScript doesn't do it for us, the idea still applies without the fancy syntax -- `while (await dequeueOrWait())` can be used instead of `for await`, and `dequeueOrWait()` can just be:

```js
async dequeueOrWait() {
  if (!this.queue.length) {
    await new Promise(resolve => this.notifyQueue = resolve);
  }
  return this.queue.unshift();
}
```

Now I think about it, it might be even easier to understand than the async iterator.

Reply:
Another idea that keeps everything inlined:

while (!this.stopped) {
  if (!this.queue.length) {
    await new Promise(resolve => this.notifyQueue = resolve);
  }
  const work = this.queue.shift();
  // ...
}

Reply:
That's really much clearer, thanks!